src/components/tl/sharp/tl_sharp_coll.c

```diff
@@ -0,0 +1,249 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "tl_sharp_coll.h"
+#include "core/ucc_mc.h"
+#include "core/ucc_ee.h"
+#include "utils/ucc_math.h"
+#include "utils/ucc_coll_utils.h"
+
+#include <sharp/api/version.h>
+#include <sharp/api/sharp_coll.h>
+
+int ucc_to_sharp_dtype[] = {
+    [UCC_DT_INT8]        = SHARP_DTYPE_NULL,
+    [UCC_DT_INT16]       = SHARP_DTYPE_SHORT,
+    [UCC_DT_INT32]       = SHARP_DTYPE_INT,
+    [UCC_DT_INT64]       = SHARP_DTYPE_LONG,
+    [UCC_DT_INT128]      = SHARP_DTYPE_NULL,
+    [UCC_DT_UINT8]       = SHARP_DTYPE_NULL,
+    [UCC_DT_UINT16]      = SHARP_DTYPE_UNSIGNED_SHORT,
+    [UCC_DT_UINT32]      = SHARP_DTYPE_UNSIGNED,
+    [UCC_DT_UINT64]      = SHARP_DTYPE_UNSIGNED_LONG,
+    [UCC_DT_UINT128]     = SHARP_DTYPE_NULL,
+    [UCC_DT_FLOAT16]     = SHARP_DTYPE_FLOAT_SHORT,
+    [UCC_DT_FLOAT32]     = SHARP_DTYPE_FLOAT,
+    [UCC_DT_FLOAT64]     = SHARP_DTYPE_DOUBLE,
+    [UCC_DT_USERDEFINED] = SHARP_DTYPE_NULL,
+    [UCC_DT_OPAQUE]      = SHARP_DTYPE_NULL,
+};
+
+int ucc_to_sharp_reduce_op[] = {
+    [UCC_OP_USERDEFINED] = SHARP_OP_NULL,
+    [UCC_OP_SUM]         = SHARP_OP_SUM,
+    [UCC_OP_PROD]        = SHARP_OP_NULL,
+    [UCC_OP_MAX]         = SHARP_OP_MAX,
+    [UCC_OP_MIN]         = SHARP_OP_MIN,
+    [UCC_OP_LAND]        = SHARP_OP_LAND,
+    [UCC_OP_LOR]         = SHARP_OP_LOR,
+    [UCC_OP_LXOR]        = SHARP_OP_LXOR,
+    [UCC_OP_BAND]        = SHARP_OP_BAND,
+    [UCC_OP_BOR]         = SHARP_OP_BOR,
+    [UCC_OP_BXOR]        = SHARP_OP_BXOR,
+    [UCC_OP_MAXLOC]      = SHARP_OP_MAXLOC,
+    [UCC_OP_MINLOC]      = SHARP_OP_MINLOC,
+};
+
+int ucc_to_sharp_memtype[] = {
+    [UCC_MEMORY_TYPE_HOST]         = SHARP_MEM_TYPE_HOST,
+    [UCC_MEMORY_TYPE_CUDA]         = SHARP_MEM_TYPE_CUDA,
+    [UCC_MEMORY_TYPE_CUDA_MANAGED] = SHARP_MEM_TYPE_LAST,
+    [UCC_MEMORY_TYPE_ROCM]         = SHARP_MEM_TYPE_LAST,
+    [UCC_MEMORY_TYPE_ROCM_MANAGED] = SHARP_MEM_TYPE_LAST,
+    [UCC_MEMORY_TYPE_LAST]         = SHARP_MEM_TYPE_LAST,
+};
+
+static ucc_tl_sharp_reg_t ucc_tl_sharp_reg_null = { .mr = NULL };
+
+static inline ucc_status_t ucc_sharp_check_dt_supported(ucc_datatype_t dt1,
+                                                        ucc_datatype_t dt2)
+{
+    if ((dt1 != dt2) || (ucc_to_sharp_dtype[dt1] == SHARP_DTYPE_NULL)) {
+        return UCC_ERR_NOT_SUPPORTED;
+    }
+
+    return UCC_OK;
+}
+
+static ucc_status_t
+ucc_tl_sharp_mem_register(ucc_tl_sharp_context_t *ctx, void *addr,
+                          size_t length, ucc_tl_sharp_reg_t **reg)
+{
+    ucc_rcache_region_t *rregion;
+    ucc_tl_sharp_rcache_region_t *region;
+    ucc_status_t status;
+    ucc_tl_sharp_reg_t *r;
+
+    if (length < ctx->cfg.reg_threshold) {
+        *reg = &ucc_tl_sharp_reg_null;
```

Comment:
just a question: if reg is NULL, does this mean libsharp internally uses "eager" buffer?

Reply:
it uses INLINE sends for small messages (default: 256 bytes), which not requires registration.

Reply:
but reg_threshold here is ,control param that is available for user. User can set it to, say, 1K which will be larger than max_inline (256). How will it be handled then?